extensions/stack_snapshot.bzl

```diff
@@ -0,0 +1,295 @@
+                module_version = module.version,
+                module_name = module.name,
+                package_name = package_name,
+            ))
+        sets.insert(packages_in_module, package_name)
+
+        # If the package is already configured, we can only only make it visible.
+        if sets.contains(conf.configured_packages, package_name):
+            if not package_tag.hidden and package_name not in conf.vendored_packages:
+                sets.insert(conf.packages, package_name)
+            continue
+        sets.insert(conf.configured_packages, package_name)
+
+        # We are the first one to configure the package:
+        if not package_tag.hidden and not package_tag.vendored:
+            sets.insert(conf.packages, package_name)
+        if package_tag.components != ["DETECT_DEFAULT"]:
+            # Some packages have default components set
+            conf.components[package_name] = package_tag.components
+        if package_tag.components_dependencies:
+            conf.components_dependencies[package_name] = json.encode(package_tag.components_dependencies)
+        if package_tag.extra_deps:
+            conf.extra_deps[package_name] = package_tag.extra_deps
+        if root_or_rules_haskell:
+            # Only the root module or rules_haskell can modify the
+            # following configuration settings. To make the configuration
+            # easier to understand.
+            if package_tag.setup_deps:
+                conf.setup_deps[package_name] = package_tag.setup_deps
+            if package_tag.flags:
+                conf.flags[package_name] = package_tag.flags
+            if package_tag.vendored:
+                conf.vendored_packages[package_name] = package_tag.vendored
+
+def _stack_snapshot_impl(mctx):
+    root_module = None
+    rules_haskell_modules = []
+    kwargs = {}
+    for module in mctx.modules:
+        if module.is_root:
+            root_module = module
+        if module.name == "rules_haskell":
+            rules_haskell_modules.append(module)
+        if module in [root_module] + rules_haskell_modules:
+            # Most modules can only add the packages they want to use to the snapshot,
+            # but only the root module can configure all the settings.
+            # rules_haskell can also set the snapshot to be used as a default.
+            if module.tags.snapshot:
+                _assert_unique_tag(module.tags.snapshot, "snapshot", module)
+                snapshot_tag = module.tags.snapshot[0]
+                if "snapshot" not in kwargs and "local_snapshot" not in kwargs:
+                    if snapshot_tag.local_snapshot:
+                        kwargs["local_snapshot"] = snapshot_tag.local_snapshot
+                    if snapshot_tag.name:
+                        kwargs["snapshot_tag"] = snapshot_tag.name
+        if module == root_module:
```

Comment:
Similarly, worth considering whether we should raise an error if a non-root module invokes these tags.